File: backend/app/schemas/monitor.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.models.monitor import Protocol


class MonitorCreate(BaseModel):
    name: str = Field(..., max_length=200)
    protocol: Protocol
    target: str = Field(..., max_length=500)
    port: int | None = None
    interval_seconds: int = 30
    timeout_seconds: float = 5.0
    is_active: bool = True
    expected_status: int | None = None
    dns_record_type: str | None = None


class MonitorUpdate(BaseModel):
    name: str | None = None
    protocol: Protocol | None = None
    target: str | None = None
    port: int | None = None
    interval_seconds: int | None = None
    timeout_seconds: float | None = None
    is_active: bool | None = None
    expected_status: int | None = None
    dns_record_type: str | None = None


class MonitorResponse(BaseModel):
    id: int
    name: str
    protocol: Protocol
    target: str
    port: int | None
    interval_seconds: int
    timeout_seconds: float
    is_active: bool
    expected_status: int | None
    dns_record_type: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/monitor.ts
export type Protocol = 'ping' | 'http' | 'https' | 'tcp' | 'dns';

export interface Monitor {
  id: number;
  name: string;
  protocol: Protocol;
  target: string;
  port: number | null;
  interval_seconds: number;
  timeout_seconds: number;
  is_active: boolean;
  expected_status: number | null;
  dns_record_type: string | null;
  created_at: string;
  updated_at: string;
}

export interface MonitorCreate {
  name: string;
  protocol: Protocol;
  target: string;
  port?: number | null;
  interval_seconds?: number;
  timeout_seconds?: number;
  is_active?: boolean;
  expected_status?: number | null;
  dns_record_type?: string | null;
}

export interface MonitorUpdate {
  name?: string;
  protocol?: Protocol;
  target?: string;
  port?: number | null;
  interval_seconds?: number;
  timeout_seconds?: number;
  is_active?: boolean;
  expected_status?: number | null;
  dns_record_type?: string | null;
}

export interface CheckResult {
  id: number;
  monitor_id: number;
  is_success: boolean;
  latency_ms: number | null;
  status_code: number | null;
  error_message: string | null;
  dns_result: string | null;
  checked_at: string;
}

export interface MonitorStats {
  monitor_id: number;
  total_checks: number;
  success_count: number;
  uptime_percent: number;
  avg_latency_ms: number | null;
  min_latency_ms: number | null;
  max_latency_ms: number | null;
  p95_latency_ms: number | null;
}
